theano/compile/builders.py

```diff
@@ -137,6 +137,46 @@ def perform(self, node, inputs, outputs):
             # we wont need this copy anymore
             output[0] = variable.copy()
 
+    def c_code(self, node, name, inputs, outputs, sub):
+        order = io_toposort(self.new_inputs, self.new_outputs)
+
+        # assert all inner nodes have c_code()
+        if not (self._op_use_c_code and
+                all([getattr(node.op, '_op_use_c_code', False) and
+                     hasattr(node.op, 'c_code') for node in order])):
+            self._op_use_c_code = False
+            return
+        else:
+            c_code = ""
+            local = locals()
+            for inner_node in order:
+                name = 'node' + str(id(inner_node))
+
+                in_names = []
+                for i, var in enumerate(inner_node.inputs):
+                    if var in self.new_inputs:
+                        in_names.append(inputs[i])
+                    else:
+                        new_name = 'InVar' + str(id(var))
+                        in_names.append(new_name)
+                        local.setdefault(new_name, var.clone())
+                        c_code += var.type.c_declare(new_name, sub)
+
+                out_names = []
+                for i, var in enumerate(inner_node.outputs):
+                    if var in self.new_outputs:
+                        out_names.append(outputs[i])
+                    else:
+                        new_name = 'OutVar' + str(id(var))
+                        out_names.append(new_name)
+                        local.setdefault(new_name, var.clone())
+                        c_code += var.type.c_declare(new_name, sub)
+
+                c_code += inner_node.op.c_code(inner_node, name,
```

Comment:
It seem it was me that started that!

https://github.com/nouiz/Theano/tree/op_from_graph

Only the 4 first commit are usefull. The last 2 are now useless.

Can you check that? Maybe there is enough comments.

On Fri, Jul 31, 2015 at 1:34 PM, abergeron <notifications@github.com> wrote:

> In theano/compile/builders.py
> <https://github.com/Theano/Theano/pull/3226#discussion_r35995984>:
>
> > +                        new_name = 'InVar' + str(id(var))
> > +                        in_names.append(new_name)
> > +                        local.setdefault(new_name, var.clone())
> > +                        c_code += var.type.c_declare(new_name, sub)
> > +
> > +                out_names = []
> > +                for i, var in enumerate(inner_node.outputs):
> > +                    if var in self.new_outputs:
> > +                        out_names.append(outputs[i])
> > +                    else:
> > +                        new_name = 'OutVar' + str(id(var))
> > +                        out_names.append(new_name)
> > +                        local.setdefault(new_name, var.clone())
> > +                        c_code += var.type.c_declare(new_name, sub)
> > +
> > +                c_code += inner_node.op.c_code(inner_node, name,
>
> I gave up before I got to the part where I would have had to compile the
> inner function of scan. So I don't have any code that would help.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3226/files#r35995984>.
>